File: java-mvc/clean-architecture/src/main/java/com/ecommerce/clean/entity/Product.java
package com.ecommerce.clean.entity;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * ENTITY LAYER (innermost) - Enterprise Business Rules.
 * Contains the most general and high-level business rules.
 * NO dependencies on any outer layer. NO framework imports.
 * This is the core of Clean Architecture.
 */
public class Product {

    private UUID id;
    private String name;
    private String description;
    private BigDecimal price;
    private String category;
    private int stockQuantity;
    private String sku;
    private boolean active;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    public Product(UUID id, String name, String description, BigDecimal price,
                   String category, int stockQuantity, String sku) {
        validate(name, price, stockQuantity);
        this.id = id != null ? id : UUID.randomUUID();
        this.name = name;
        this.description = description;
        this.price = price;
        this.category = category;
        this.stockQuantity = stockQuantity;
        this.sku = sku;
        this.active = true;
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    // Business rules embedded in the entity
    public void updateDetails(String name, String description, BigDecimal price, String category) {
        validate(name, price, this.stockQuantity);
        this.name = name;
        this.description = description;
        this.price = price;
        this.category = category;
        this.updatedAt = LocalDateTime.now();
    }

    public void decreaseStock(int quantity) {
        if (quantity <= 0) throw new IllegalArgumentException("Quantity must be positive");
        if (stockQuantity < quantity) throw new IllegalStateException(
                "Insufficient stock. Available: " + stockQuantity + ", Requested: " + quantity);
        this.stockQuantity -= quantity;
        this.updatedAt = LocalDateTime.now();
    }

    public void increaseStock(int quantity) {
        if (quantity <= 0) throw new IllegalArgumentException("Quantity must be positive");
        this.stockQuantity += quantity;
        this.updatedAt = LocalDateTime.now();
    }

    public boolean isInStock() {
        return stockQuantity > 0 && active;
    }

    public void deactivate() {
        this.active = false;
        this.updatedAt = LocalDateTime.now();
    }

    private void validate(String name, BigDecimal price, int stock) {
        if (name == null || name.isBlank()) throw new IllegalArgumentException("Name is required");
        if (name.length() > 255) throw new IllegalArgumentException("Name too long");
        if (price == null || price.compareTo(BigDecimal.ZERO) <= 0)
            throw new IllegalArgumentException("Price must be positive");
        if (stock < 0) throw new IllegalArgumentException("Stock cannot be negative");
    }

    // Getters
    public UUID getId() { return id; }
    public String getName() { return name; }
    public String getDescription() { return description; }
    public BigDecimal getPrice() { return price; }
    public String getCategory() { return category; }
    public int getStockQuantity() { return stockQuantity; }
    public String getSku() { return sku; }
    public boolean isActive() { return active; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }

    // For reconstitution from persistence
    public void setId(UUID id) { this.id = id; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    public void setActive(boolean active) { this.active = active; }
}

File: dotnet-mvc/CleanArchitecture/Entities/Product.cs
namespace CleanArchitecture.Entities;

/// <summary>
/// Enterprise Business Entity — the innermost layer of Clean Architecture.
/// Contains enterprise-wide business rules that would exist even if the system were manual.
/// This entity has NO dependencies on any outer layer (use cases, adapters, or frameworks).
/// </summary>
public class Product
{
    public Guid Id { get; private set; }
    public string Name { get; private set; } = string.Empty;
    public string Description { get; private set; } = string.Empty;
    public decimal Price { get; private set; }
    public string Category { get; private set; } = string.Empty;
    public int StockQuantity { get; private set; }
    public string Sku { get; private set; } = string.Empty;
    public bool Active { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime UpdatedAt { get; private set; }

    private Product() { }

    /// <summary>
    /// Factory method enforcing enterprise business rules for product creation.
    /// </summary>
    public static Product Create(string name, string description, decimal price, string category, int stockQuantity, string sku)
    {
        ValidateName(name);
        ValidatePrice(price);
        ValidateSku(sku);
        ValidateStock(stockQuantity);

        return new Product
        {
            Id = Guid.NewGuid(),
            Name = name,
            Description = description ?? string.Empty,
            Price = price,
            Category = category ?? string.Empty,
            StockQuantity = stockQuantity,
            Sku = sku,
            Active = true,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
    }

    /// <summary>
    /// Reconstitutes a product from persistence (used by gateways).
    /// </summary>
    public static Product Reconstitute(Guid id, string name, string description, decimal price,
        string category, int stockQuantity, string sku, bool active, DateTime createdAt, DateTime updatedAt)
    {
        return new Product
        {
            Id = id,
            Name = name,
            Description = description,
            Price = price,
            Category = category,
            StockQuantity = stockQuantity,
            Sku = sku,
            Active = active,
            CreatedAt = createdAt,
            UpdatedAt = updatedAt
        };
    }

    /// <summary>
    /// Enterprise rule: update product attributes.
    /// </summary>
    public void Update(string name, string description, decimal price, string category, string sku)
    {
        ValidateName(name);
        ValidatePrice(price);

        Name = name;
        Description = description ?? string.Empty;
        Price = price;
        Category = category ?? string.Empty;
        Sku = sku;
        UpdatedAt = DateTime.UtcNow;
    }

    /// <summary>
    /// Enterprise rule: stock cannot go below zero.
    /// </summary>
    public void DecreaseStock(int quantity)
    {
        if (quantity <= 0)
            throw new EntityValidationException("Quantity must be greater than zero.");

        if (StockQuantity - quantity < 0)
            throw new EntityValidationException("Insufficient stock. Cannot decrease below zero.");

        StockQuantity -= quantity;
        UpdatedAt = DateTime.UtcNow;
    }

    /// <summary>
    /// Enterprise rule: increase stock.
    /// </summary>
    public void IncreaseStock(int quantity)
    {
        if (quantity <= 0)
            throw new EntityValidationException("Quantity must be greater than zero.");

        StockQuantity += quantity;
        UpdatedAt = DateTime.UtcNow;
    }

    /// <summary>
    /// Enterprise rule: soft delete.
    /// </summary>
    public void Deactivate()
    {
        Active = false;
        UpdatedAt = DateTime.UtcNow;
    }

    private static void ValidateName(string name)
    {
        if (string.IsNullOrWhiteSpace(name))
            throw new EntityValidationException("Product name is required.");
    }

    private static void ValidatePrice(decimal price)
    {
        if (price <= 0)
            throw new EntityValidationException("Price must be greater than zero.");
    }

    private static void ValidateSku(string sku)
    {
        if (string.IsNullOrWhiteSpace(sku))
            throw new EntityValidationException("SKU is required.");
    }

    private static void ValidateStock(int stock)
    {
        if (stock < 0)
            throw new EntityValidationException("Stock quantity cannot be negative.");
    }
}

/// <summary>
/// Exception for entity-level validation failures (enterprise business rules).
/// </summary>
public class EntityValidationException : Exception
{
    public EntityValidationException(string message) : base(message) { }
}
